subiquity/common/filesystem/tests/test_manipulator.py

```diff
@@ -234,3 +236,201 @@ def test_add_boot_has_valid_offset(self):
+
+    def test_add_boot_UEFI_full_resizes_larger(self):
+        manipulator = make_manipulator(Bootloader.UEFI)
+        # 402MiB so that the space available for partitioning (400MiB)
+        # divided by 4 is an whole number of megabytes.
+        disk = make_disk(manipulator.model, preserve=True, size=402*MiB)
+        part_smaller = make_partition(
+            manipulator.model, disk, size=gaps.largest_gap_size(disk)//4)
+        part_larger = make_partition(
+            manipulator.model, disk, size=gaps.largest_gap_size(disk))
+        larger_size_before = part_larger.size
+        smaller_size_before = part_smaller.size
+        manipulator.add_boot_disk(disk)
+        self.assertIsBootDisk(manipulator, disk)
+        [p1, p2, p3] = sorted(disk.partitions(), key=lambda p: p.offset)
+        self.assertIs(p1, part_smaller)
+        self.assertIs(p3, part_larger)
+        self.assertEqual(smaller_size_before, p1.size)
+        self.assertEqual(p1.offset, MiB)
+        self.assertEqual(p2.offset, p1.offset + p1.size)
+        self.assertEqual(p3.offset, p2.offset + p2.size)
+        self.assertTrue(boot.is_esp(p2))
+        self.assertEqual(p3.size, larger_size_before - p2.size)
+
+    def test_add_boot_PREP_empty(self):
+        manipulator = make_manipulator(Bootloader.PREP)
+        disk = make_disk(manipulator.model, preserve=True)
+        manipulator.add_boot_disk(disk)
+        self.assertIsBootDisk(manipulator, disk)
+        [part] = disk.partitions()
+        self.assertEqual(part.offset, MiB)
+
+    def test_add_boot_PREP_full(self):
+        manipulator = make_manipulator(Bootloader.PREP)
+        disk = make_disk(manipulator.model, preserve=True)
+        part = make_partition(
+            manipulator.model, disk, size=gaps.largest_gap_size(disk))
+        size_before = part.size
+        manipulator.add_boot_disk(disk)
+        self.assertIsBootDisk(manipulator, disk)
+        [p1, p2] = disk.partitions()
+        self.assertIs(p2, part)
+        size_after = p2.size
+        self.assertEqual(p1.offset, MiB)
+        self.assertEqual(p2.offset, MiB+p1.size)
+        self.assertEqual(size_after, size_before - p1.size)
+
+    def test_add_boot_PREP_half_full(self):
+        manipulator = make_manipulator(Bootloader.PREP)
+        disk = make_disk(manipulator.model, preserve=True)
+        part = make_partition(
+            manipulator.model, disk, size=gaps.largest_gap_size(disk)//2)
+        size_before = part.size
+        manipulator.add_boot_disk(disk)
+        self.assertIsBootDisk(manipulator, disk)
+        [p1, p2] = sorted(disk.partitions(), key=lambda p: p.offset)
+        size_after = p1.size
+        self.assertIs(p1, part)
+        self.assertEqual(p1.offset, MiB)
+        self.assertEqual(p2.offset, p1.offset + p1.size)
+        self.assertEqual(p2.flag, "prep")
+        self.assertEqual(size_after, size_before)
+
+    def test_add_boot_PREP_full_resizes_larger(self):
```

Comment:
Obviously there are some similarities between the different bootloader tests.
What your thoughts on refactoring them a bit?

Reply:
Yeah they are pretty repetitive. I wonder if some well chosen utility functions might help... not today though.